src/main/scala/habla/doric/Errors.scala

```diff
@@ -4,11 +4,72 @@ import cats.data.NonEmptyChain
 
 import org.apache.spark.sql.types.DataType
 
+object ErrorUtils {
+  def getSingularOrPlural(nErrors: Long): String =
+    if (nErrors > 1) "errors" else "error"
+}
+
+case class DoricJoinMultiError(
+    errors: NonEmptyChain[JoinDoricSingleError]
+) extends Throwable(
+      errors.head.getCause
+    ) {
+  override def getMessage: String = {
+    implicit class StringOps(s: String) {
+      def addTabs: String = s.replaceAll("\n", "\n\t")
+    }
+
+    implicit class DoricMulti(multiError: DoricMultiError) {
+      def getForSide(side: String): String =
+        s"$side dataframe:\n${multiError.getMessage.addTabs}"
+    }
+
+    val numErrors: Long = errors.length
+    val leftErrors = NonEmptyChain
+      .fromChain(errors.toChain.filter(_.isLeft))
+      .map(DoricMultiError(_).getForSide("Left"))
+    val rightErrors = NonEmptyChain
+      .fromChain(errors.toChain.filter(_.isRight))
+      .map(DoricMultiError(_).getForSide("Right"))
+
+    val sidesErrors = (leftErrors, rightErrors) match {
+      case (Some(l), Some(r)) => l + "\n" + r
+      case (Some(l), _)       => l
+      case (_, Some(r))       => r
+      case _                  => "impossible!!!"
```

Comment:
The impossible case is annoying, is it caused by avoiding `Either[A, Either[A, (A, A)]]`?

Reply:
the tuple is created in this function, and can't be accessed outside, it's a private annoying thing to not have a warning of not exhaustive matching

Reply:
The public type is the `Validated[DoricJoinError, T]`

Reply:
Changed to a less confusing way to create the string